src/Controls/src/BindingSourceGen/BindingCodeWriter.cs

```diff
@@ -274,20 +274,21 @@ private void AppendSetterAction(SetBindingInvocationDescription binding, string
 			}
 		}
 
-		private void AppendHandlersArray(TypeDescription sourceType, EquatableArray<IPathPart> path)
+		private void AppendHandlersArray(TypeDescription sourceType, EquatableArray<IPathPart> path, bool considerAllReferenceTypesPotentiallyNullable)
 		{
 			AppendLine($"new Tuple<Func<{sourceType}, object?>, string>[]");
 			AppendLine('{');
 
 			Indent();
 
 			string nextExpression = "source";
-			bool forceConditonalAccessToNextPart = false;
+			bool forceConditonalAccessToNextPart = !sourceType.IsValueType && considerAllReferenceTypesPotentiallyNullable;
 			foreach (var part in path)
 			{
 				var previousExpression = nextExpression;
 				nextExpression = AccessExpressionBuilder.Build(previousExpression, MaybeWrapInConditionalAccess(part, forceConditonalAccessToNextPart));
-				forceConditonalAccessToNextPart = part is Cast;
+				var isNullableReferenceType = part is MemberAccess memberAccess && !memberAccess.IsValueType;
```

Comment:
We need a check for `IndexAccess` here as well + a unit test that covers that case.